File: sdk/c/wasmify.c
/*
 * Wasmify C SDK Implementation
 */

#include "wasmify.h"
#include <time.h>

// Global initialization state
static int g_initialized = 0;

// HTTP response callback
static size_t write_callback(void* contents, size_t size, size_t nmemb, void* userp) {
    size_t realsize = size * nmemb;
    wasmify_response_t* response = (wasmify_response_t*)userp;
    
    char* new_data = realloc(response->data, response->size + realsize + 1);
    if (!new_data) {
        return 0;
    }
    
    response->data = new_data;
    memcpy(&(response->data[response->size]), contents, realsize);
    response->size += realsize;
    response->data[response->size] = 0;
    
    return realsize;
}

// Initialize Wasmify SDK
wasmify_error_t wasmify_init(void) {
    if (g_initialized) {
        return WASMIFY_SUCCESS;
    }
    
    curl_global_init(CURL_GLOBAL_DEFAULT);
    g_initialized = 1;
    
    return WASMIFY_SUCCESS;
}

// Cleanup Wasmify SDK
void wasmify_cleanup(void) {
    if (g_initialized) {
        curl_global_cleanup();
        g_initialized = 0;
    }
}

// Create a new Wasmify client
wasmify_client_t* wasmify_client_create(wasmify_config_t config) {
    wasmify_client_t* client = malloc(sizeof(wasmify_client_t));
    if (!client) {
        return NULL;
    }
    
    // Copy configuration
    client->config.api_url = config.api_url ? strdup(config.api_url) : strdup("http://localhost:3000/api");
    client->config.api_key = config.api_key ? strdup(config.api_key) : NULL;
    client->config.timeout = config.timeout > 0 ? config.timeout : 30;
    
    // Initialize CURL
    client->curl = curl_easy_init();
    if (!client->curl) {
        free(client->config.api_url);
        if (client->config.api_key) free(client->config.api_key);
        free(client);
        return NULL;
    }
    
    // Set common CURL options
    curl_easy_setopt(client->curl, CURLOPT_TIMEOUT, client->config.timeout);
    curl_easy_setopt(client->curl, CURLOPT_FOLLOWLOCATION, 1L);
    curl_easy_setopt(client->curl, CURLOPT_WRITEFUNCTION, write_callback);
    
    return client;
}

// Destroy a Wasmify client
void wasmify_client_destroy(wasmify_client_t* client) {
    if (!client) return;
    
    if (client->curl) {
        curl_easy_cleanup(client->curl);
    }
    if (client->config.api_url) {
        free(client->config.api_url);
    }
    if (client->config.api_key) {
        free(client->config.api_key);
    }
    free(client);
}

// Execute HTTP request
static wasmify_error_t execute_request(
    wasmify_client_t* client,
    const char* url,
    const char* post_data,
    wasmify_response_t* response
) {
    if (!client || !url || !response) {
        return WASMIFY_ERROR_INVALID_PARAM;
    }
    
    // Initialize response
    response->data = malloc(1);
    response->data[0] = '\0';
    response->size = 0;
    
    // Set URL
    curl_easy_setopt(client->curl, CURLOPT_URL, url);
    
    // Set POST data if provided
    if (post_data) {
        curl_easy_setopt(client->curl, CURLOPT_POSTFIELDS, post_data);
        curl_easy_setopt(client->curl, CURLOPT_POSTFIELDSIZE, strlen(post_data));
    }
    
    // Set response callback
    curl_easy_setopt(client->curl, CURLOPT_WRITEDATA, response);
    
    // Set headers
    struct curl_slist* headers = NULL;
    headers = curl_slist_append(headers, "Content-Type: application/json");
    if (client->config.api_key) {
        char auth_header[256];
        snprintf(auth_header, sizeof(auth_header), "Authorization: Bearer %s", client->config.api_key);
        headers = curl_slist_append(headers, auth_header);
    }
    curl_easy_setopt(client->curl, CURLOPT_HTTPHEADER, headers);
    
    // Execute request
    CURLcode res = curl_easy_perform(client->curl);
    
    // Cleanup headers
    curl_slist_free_all(headers);
    
    if (res != CURLE_OK) {
        free(response->data);
        return WASMIFY_ERROR_NETWORK;
    }
    
    // Check HTTP response code
    long response_code;
    curl_easy_getinfo(client->curl, CURLINFO_RESPONSE_CODE, &response_code);
    
    if (response_code != 200) {
        free(response->data);
        return WASMIFY_ERROR_NETWORK;
    }
    
    return WASMIFY_SUCCESS;
}

// Upload a WebAssembly module
wasmify_error_t wasmify_upload_module(
    wasmify_client_t* client,
    const char* file_path,
    const char* name,
    const char* version,
    wasmify_module_t* module
) {
    if (!client || !file_path || !name || !version || !module) {
        return WASMIFY_ERROR_INVALID_PARAM;
    }
    
    // For now, simulate upload
    module->id = strdup("simulated-module-id");
    module->name = strdup(name);
    module->version = strdup(version);
    module->file_path = strdup(file_path);
    module->metadata = cJSON_CreateObject();
    
    return WASMIFY_SUCCESS;
}

// Execute a WebAssembly module function
wasmify_error_t wasmify_execute_module(
    wasmify_client_t* client,
    const char* module_id,
    const char* function_name,
    char** args,
    int args_count,
    wasmify_result_t* result
) {
    if (!client || !module_id || !function_name || !result) {
        return WASMIFY_ERROR_INVALID_PARAM;
    }
    
    // Create JSON request
    cJSON* request = cJSON_CreateObject();
    cJSON* config = cJSON_CreateObject();
    cJSON* memory = cJSON_CreateObject();
    
    cJSON_AddStringToObject(request, "moduleId", module_id);
    cJSON_AddStringToObject(request, "functionName", function_name);
    
    // Add args array
    cJSON* args_array = cJSON_CreateArray();
    for (int i = 0; i < args_count; i++) {
        cJSON_AddItemToArray(args_array, cJSON_CreateString(args[i]));
    }
    cJSON_AddItemToObject(request, "args", args_array);
    
    // Add config
    cJSON_AddNumberToObject(memory, "min", 64);
    cJSON_AddNumberToObject(memory, "max", 512);
    cJSON_AddItemToObject(config, "memory", memory);
    cJSON_AddNumberToObject(config, "maxExecutionTime", 30000);
    cJSON_AddBoolToObject(config, "enableWasi", 1);
    cJSON_AddItemToObject(request, "config", config);
    
    char* json_data = cJSON_Print(request);
    
    // Build URL
    char url[512];
    snprintf(url, sizeof(url), "%s/wasm/execute", client->config.api_url);
    
    // Execute request
    wasmify_response_t response;
    wasmify_error_t error = execute_request(client, url, json_data, NULL);
    
    free(json_data);
    cJSON_Delete(request);
    
    if (error != WASMIFY_SUCCESS) {
        return error;
    }
    
    // Simulate result
    result->success = 1;
    result->result = strdup("simulated execution result");
    result->execution_time = 42.5;
    result->memory_used = 1024 * 1024;
    result->error = NULL;
    
    return WASMIFY_SUCCESS;
}

// Execute WebAssembly module locally
wasmify_error_t wasmify_execute_local(
    const char* file_path,
    const char* function_name,
    char** args,
    int args_count,
    wasmify_result_t* result
) {
    if (!file_path || !function_name || !result) {
        return WASMIFY_ERROR_INVALID_PARAM;
    }
    
    // Simulate local execution
    clock_t start = clock();
    
    // Simulate execution time
    for (volatile int i = 0; i < 1000000; i++);
    
    double execution_time = ((double)(clock() - start)) / CLOCKS_PER_SEC * 1000;
    
    // Create result
    result->success = 1;
    result->result = malloc(256);
    snprintf(result->result, 256, "Executed %s with %d args", function_name, args_count);
    result->execution_time = execution_time;
    result->memory_used = 1024 * 1024; // 1MB
    result->error = NULL;
    
    return WASMIFY_SUCCESS;
}

// List all available modules
wasmify_error_t wasmify_list_modules(
    wasmify_client_t* client,
    wasmify_module_t** modules,
    int* modules_count
) {
    if (!client || !modules || !modules_count) {
        return WASMIFY_ERROR_INVALID_PARAM;
    }
    
    // For now, return empty list
    *modules = NULL;
    *modules_count = 0;
    
    return WASMIFY_SUCCESS;
}

// Deploy module to edge locations
wasmify_error_t wasmify_deploy_to_edge(
    wasmify_client_t* client,
    const char* module_id,
    char** regions,
    int regions_count,
    char** deployment_id
) {
    if (!client || !module_id || !deployment_id) {
        return WASMIFY_ERROR_INVALID_PARAM;
    }
    
    // Simulate deployment
    *deployment_id = strdup("simulated-deployment-id");
    
    return WASMIFY_SUCCESS;
}

// Free module structure
void wasmify_module_free(wasmify_module_t* module) {
    if (!module) return;
    
    if (module->id) free(module->id);
    if (module->name) free(module->name);
    if (module->version) free(module->version);
    if (module->file_path) free(module->file_path);
    if (module->metadata) cJSON_Delete(module->metadata);
    free(module);
}

// Free result structure
void wasmify_result_free(wasmify_result_t* result) {
    if (!result) return;
    
    if (result->result) free(result->result);
    if (result->error) free(result->error);
    free(result);
}

// Convenience function to run WASM quickly
wasmify_error_t wasmify_run(
    const char* file_path,
    const char* function_name,
    char** args,
    int args_count,
    char** output
) {
    wasmify_result_t result;
    wasmify_error_t error = wasmify_execute_local(file_path, function_name, args, args_count, &result);
    
    if (error == WASMIFY_SUCCESS && result.success) {
        *output = strdup(result.result);
    }
    
    wasmify_result_free(&result);
    return error;
}

File: sdk/go/wasmify.go
package wasmify

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"mime/multipart"
	"net/http"
	"os"
	"path/filepath"
	"time"
)

// WasmModule represents a WebAssembly module
type WasmModule struct {
	ID       string                 `json:"id"`
	Name     string                 `json:"name"`
	Version  string                 `json:"version"`
	FilePath string                 `json:"filePath"`
	Metadata map[string]interface{} `json:"metadata"`
}

// ExecutionResult represents the result of WebAssembly execution
type ExecutionResult struct {
	Success       bool        `json:"success"`
	Result        interface{} `json:"result"`
	ExecutionTime float64     `json:"executionTime"`
	MemoryUsed    int64       `json:"memoryUsed"`
	Error         string      `json:"error,omitempty"`
}

// Config represents client configuration
type Config struct {
	APIURL  string
	APIKey  string
	Timeout time.Duration
}

// Client represents the Wasmify Go client
type Client struct {
	config    Config
	httpClient *http.Client
}

// NewClient creates a new Wasmify client
func NewClient(config Config) *Client {
	if config.Timeout == 0 {
		config.Timeout = 30 * time.Second
	}
	
	return &Client{
		config: config,
		httpClient: &http.Client{
			Timeout: config.Timeout,
		},
	}
}

// NewDefaultClient creates a client with default configuration
func NewDefaultClient() *Client {
	return NewClient(Config{
		APIURL: "http://localhost:3000/api",
	})
}

// UploadModule uploads a WebAssembly module to Wasmify
func (c *Client) UploadModule(filePath, name, version string) (*WasmModule, error) {
	file, err := os.Open(filePath)
	if err != nil {
		return nil, fmt.Errorf("failed to open file: %w", err)
	}
	defer file.Close()

	// Create multipart form
	var requestBody bytes.Buffer
	writer := multipart.NewWriter(&requestBody)

	// Add file
	part, err := writer.CreateFormFile("file", filepath.Base(filePath))
	if err != nil {
		return nil, fmt.Errorf("failed to create form file: %w", err)
	}
	
	_, err = io.Copy(part, file)
	if err != nil {
		return nil, fmt.Errorf("failed to copy file: %w", err)
	}

	// Add form fields
	_ = writer.WriteField("name", name)
	_ = writer.WriteField("version", version)
	
	err = writer.Close()
	if err != nil {
		return nil, fmt.Errorf("failed to close writer: %w", err)
	}

	// Create request
	req, err := http.NewRequest("POST", c.config.APIURL+"/upload", &requestBody)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Content-Type", writer.FormDataContentType())
	if c.config.APIKey != "" {
		req.Header.Set("Authorization", "Bearer "+c.config.APIKey)
	}

	// Send request
	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to send request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("upload failed with status: %s", resp.Status)
	}

	// Parse response
	var result struct {
		Success bool `json:"success"`
		Data    struct {
			Key     string                 `json:"key"`
			ETag    string                 `json:"etag"`
			Size    int64                  `json:"size"`
			Headers map[string]interface{} `json:"headers"`
		} `json:"data"`
	}

	err = json.NewDecoder(resp.Body).Decode(&result)
	if err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	if !result.Success {
		return nil, fmt.Errorf("upload failed")
	}

	return &WasmModule{
		ID:       result.Data.Key,
		Name:     name,
		Version:  version,
		FilePath: filePath,
		Metadata: map[string]interface{}{
			"etag":    result.Data.ETag,
			"size":    result.Data.Size,
			"headers": result.Data.Headers,
		},
	}, nil
}

// ExecuteModule executes a WebAssembly module function
func (c *Client) ExecuteModule(moduleID, functionName string, args []interface{}, config map[string]interface{}) (*ExecutionResult, error) {
	requestData := map[string]interface{}{
		"moduleId":     moduleID,
		"functionName": functionName,
		"args":         args,
		"config": map[string]interface{}{
			"memory":          map[string]int{"min": 64, "max": 512},
			"maxExecutionTime": 30000,
			"enableWasi":      true,
		},
	}

	// Merge user config
	for k, v := range config {
		requestData["config"].(map[string]interface{})[k] = v
	}

	jsonData, err := json.Marshal(requestData)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal request: %w", err)
	}

	req, err := http.NewRequest("POST", c.config.APIURL+"/wasm/execute", bytes.NewBuffer(jsonData))
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Content-Type", "application/json")
	if c.config.APIKey != "" {
		req.Header.Set("Authorization", "Bearer "+c.config.APIKey)
	}

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to send request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("execution failed with status: %s", resp.Status)
	}

	var result struct {
		Success bool `json:"success"`
		Data    struct {
			Result struct {
				Result        interface{} `json:"result"`
				ExecutionTime float64     `json:"executionTime"`
				MemoryUsed    int64       `json:"memoryUsed"`
				Error         string      `json:"error,omitempty"`
			} `json:"result"`
		} `json:"data"`
	}

	err = json.NewDecoder(resp.Body).Decode(&result)
	if err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	if !result.Success {
		return nil, fmt.Errorf("execution failed")
	}

	return &ExecutionResult{
		Success:       true,
		Result:        result.Data.Result.Result,
		ExecutionTime: result.Data.Result.ExecutionTime,
		MemoryUsed:    result.Data.Result.MemoryUsed,
		Error:         result.Data.Result.Error,
	}, nil
}

// ListModules lists all available WebAssembly modules
func (c *Client) ListModules() ([]*WasmModule, error) {
	req, err := http.NewRequest("GET", c.config.APIURL+"/modules", nil)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	if c.config.APIKey != "" {
		req.Header.Set("Authorization", "Bearer "+c.config.APIKey)
	}

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to send request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("list failed with status: %s", resp.Status)
	}

	var result struct {
		Success bool `json:"success"`
		Data    []struct {
			ID          string                 `json:"id"`
			Name        string                 `json:"name"`
			Version     string                 `json:"version"`
			WasmFile    string                 `json:"wasmFile"`
			Description string                 `json:"description"`
			Language    string                 `json:"language"`
			Size        int64                  `json:"size"`
			Hash        string                 `json:"hash"`
			IsPublic    bool                   `json:"isPublic"`
			CreatedAt   string                 `json:"createdAt"`
			UpdatedAt   string                 `json:"updatedAt"`
		} `json:"data"`
	}

	err = json.NewDecoder(resp.Body).Decode(&result)
	if err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	if !result.Success {
		return nil, fmt.Errorf("list failed")
	}

	modules := make([]*WasmModule, len(result.Data))
	for i, moduleData := range result.Data {
		modules[i] = &WasmModule{
			ID:       moduleData.ID,
			Name:     moduleData.Name,
			Version:  moduleData.Version,
			FilePath: moduleData.WasmFile,
			Metadata: map[string]interface{}{
				"description": moduleData.Description,
				"language":    moduleData.Language,
				"size":        moduleData.Size,
				"hash":        moduleData.Hash,
				"isPublic":    moduleData.IsPublic,
				"createdAt":   moduleData.CreatedAt,
				"updatedAt":   moduleData.UpdatedAt,
			},
		}
	}

	return modules, nil
}

// DeployToEdge deploys a module to edge locations
func (c *Client) DeployToEdge(moduleID string, regions []string) (map[string]interface{}, error) {
	requestData := map[string]interface{}{
		"moduleId":    moduleID,
		"environment": "production",
		"region":      "global",
		"config": map[string]interface{}{
			"memory":   "128MB",
			"cpu":      "100m",
			"replicas": 3,
			"edge":     true,
		},
	}

	if len(regions) > 0 {
		requestData["region"] = regions[0]
	}

	jsonData, err := json.Marshal(requestData)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal request: %w", err)
	}

	req, err := http.NewRequest("POST", c.config.APIURL+"/deployments", bytes.NewBuffer(jsonData))
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Content-Type", "application/json")
	if c.config.APIKey != "" {
		req.Header.Set("Authorization", "Bearer "+c.config.APIKey)
	}

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to send request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("deployment failed with status: %s", resp.Status)
	}

	var result struct {
		Success bool                   `json:"success"`
		Data    map[string]interface{} `json:"data"`
	}

	err = json.NewDecoder(resp.Body).Decode(&result)
	if err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	if !result.Success {
		return nil, fmt.Errorf("deployment failed")
	}

	return result.Data, nil
}

// ExecuteLocal executes WebAssembly module locally (simulated)
func ExecuteLocal(wasmFilePath, functionName string, args []interface{}) (*ExecutionResult, error) {
	// This would integrate with Wasmtime Go bindings
	// For now, we'll simulate local execution
	startTime := time.Now()
	
	// Simulate execution
	result := fmt.Sprintf("Executed %s with args %v", functionName, args)
	executionTime := time.Since(startTime).Seconds() * 1000

	return &ExecutionResult{
		Success:       true,
		Result:        result,
		ExecutionTime: executionTime,
		MemoryUsed:    1024 * 1024, // 1MB
		Error:         "",
	}, nil
}

// Convenience functions
func RunWasm(wasmFilePath, functionName string, args []interface{}) (interface{}, error) {
	result, err := ExecuteLocal(wasmFilePath, functionName, args)
	if err != nil {
		return nil, err
	}

	if !result.Success {
		return nil, fmt.Errorf("WebAssembly execution failed: %s", result.Error)
	}

	return result.Result, nil
}

func DeployToCloud(wasmFilePath, name string, regions []string) (string, error) {
	client := NewDefaultClient()
	
	// Upload module
	module, err := client.UploadModule(wasmFilePath, name, "1.0.0")
	if err != nil {
		return "", err
	}

	// Deploy to edge
	deployment, err := client.DeployToEdge(module.ID, regions)
	if err != nil {
		return "", err
	}

	return deployment["id"].(string), nil
}